handlers/mist_triggers.go

```diff
@@ -0,0 +1,207 @@
+package handlers
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/julienschmidt/httprouter"
+	"github.com/livepeer/catalyst-api/clients"
+	"github.com/livepeer/catalyst-api/errors"
+)
+
+type MistCallbackHandlersCollection struct {
+	MistClient  MistAPIClient
+	StreamCache *StreamCache
+}
+
+// Trigger dispatches request to mapped method according to trigger name
+// Only single trigger callback is allowed on Mist.
+// All created streams and our handlers (segmenting, transcoding, et.) must share this endpoint.
+// If handler logic grows more complicated we may consider adding dispatch mechanism here.
+func (d *MistCallbackHandlersCollection) Trigger() httprouter.Handle {
+	return func(w http.ResponseWriter, req *http.Request, params httprouter.Params) {
+		t := &Trigger{w: w, req: req}
+		whichOne := req.Header.Get("X-Trigger")
+		switch whichOne {
+		case "PUSH_END":
+			d.TriggerPushEnd(t)
+		case "LIVE_TRACK_LIST":
+			d.TriggerLiveTrackList(t)
+		default:
+			errors.WriteHTTPBadRequest(w, "Unsupported X-Trigger", fmt.Errorf("unknown trigger '%s'", whichOne))
+			return
+		}
+	}
+}
+
+// TriggerLiveTrackList responds to LIVE_TRACK_LIST trigger.
+// It is stream-specific and must be blocking. The payload for this trigger is multiple lines,
+// each separated by a single newline character (without an ending newline), containing data:
+//   stream name
+//   push target URI
+// TriggerLiveTrackList is used only by transcoding.
+func (d *MistCallbackHandlersCollection) TriggerLiveTrackList(t *Trigger) {
+	payload, err := io.ReadAll(t.req.Body)
+	if err != nil {
+		t.Err("Cannot read payload", err)
+		return
+	}
+	lines := strings.Split(strings.TrimSuffix(string(payload), "\n"), "\n")
+	t.streamName = lines[0]
+	encodedTracks := lines[1]
+
+	yes, suffix := isTranscodeStream(t.streamName)
+	if !yes {
+		t.Err("unknown streamName format", nil)
+		return
+	}
+
+	if streamEnded := encodedTracks == "null"; streamEnded {
+		// SOURCE_PREFIX stream is no longer needed
+		inputStream := fmt.Sprintf("%s%s", SOURCE_PREFIX, suffix)
+		if err = d.MistClient.DeleteStream(inputStream); err != nil {
+			log.Printf("ERROR LIVE_TRACK_LIST DeleteStream(%s) %v", inputStream, err)
+		}
+		// Multiple pushes from RENDITION_PREFIX are in progress.
+		return
+	}
+	tracks := make(LiveTrackListTriggerJson)
+	if err = json.Unmarshal([]byte(encodedTracks), &tracks); err != nil {
+		t.Err("LiveTrackListTriggerJson json decode error", err)
+		return
+	}
+	// Start push per each video track
+	info, err := d.StreamCache.Transcoding.Get(t.streamName)
+	if err != nil {
+		t.Err("LIVE_TRACK_LIST unknown push source", err)
+		return
+	}
+	for i := range tracks { // i is generated name, not important, all info is contained in element
+		if tracks[i].Type != "video" {
+			// Only produce an rendition per each video track, selecting best audio track
```

Comment:
I don't get this comment. So the other `Type` can be `Audio`, right? And then why don't we need to process it?

Reply:
The process looks as follows:
1. Add SOURCE stream to Mist using MistClient
2. Execute the `MistProceLivepeer` process to perform the following operations:
   - Take SOURCE stream and segment it
   - Push each segment to Broadcaster
   - Receive transcoded segment from Broadcaster, ingest into RENDITION stream
3. Start SOURCE stream by creating, unused, push to /dev/null.ts
4. `MistProceLivepeer` process is unblocked, starts to work
5. Return 200 OK response
5. Respond to LIVE_TRACK_LIST trigger:
   - Extract track information
   - **> For each video track create push to S3 destination**
   - Transcoding is complete if there is no tracks. Then delete SOURCE stream
6. Respond to PUSH_END trigger:
   - Determine is it error or successful upload
   - Invoke callback to studio
7. When all started pushes are complete invoke callback to studio; remove stream from StreamCache

We are at the step: For each video track create push to S3 destination

Note that we want to create self-contained rendition segments, with both one audio and one video track. Our RENDITION stream contains all video tracks and duplicate audio tracks.

Finally for each video track we start new push to create separate rendition. If we had `x` transcoded video tracks from `broadcaster` we want `x` new pushes to produce `x` `mpegts` segments on S3 storage.